File: evaluation/S0/python/Hard/1814-remove-invalid-parentheses.py
# Source: https://leetcode.com/problems/remove-invalid-parentheses/   |   Difficulty: Hard
#
# Problem Description:
# Given a string s that contains parentheses and letters, remove the minimum number of invalid parentheses to make the input string valid.
#
# Return a list of unique strings that are valid with the minimum number of removals. You may return the answer in any order.
#
# Example:
# Input: s = "()())()"
# Output: ["(())()","()()()"]
#
# Constraints:
# 1 <= s.length <= 25
# 	s consists of lowercase English letters and parentheses '(' and ')'.
# 	There will be at most 20 parentheses in s.
#

class Solution(object):
    def removeInvalidParentheses(self, s):
        """
        :type s: str
        :rtype: List[str]
        """
        res = []
        level = {s}
        
        while True:
            valid = list(filter(self.isValid, level))
            if valid:
                return valid
            
            next_level = set()
            for item in level:
                for i in range(len(item)):
                    if item[i] in ('(', ')'):
                        next_level.add(item[:i] + item[i+1:])
            level = next_level  

File: evaluation/S0/cpp/Hard/1814-remove-invalid-parentheses.cpp
// Source: https://leetcode.com/problems/remove-invalid-parentheses/   |   Difficulty: Hard
//
// Problem Description:
// Given a string s that contains parentheses and letters, remove the minimum number of invalid parentheses to make the input string valid.
//
// Return a list of unique strings that are valid with the minimum number of removals. You may return the answer in any order.
//
// Example:
// Input: s = "()())()"
// Output: ["(())()","()()()"]
//
// Constraints:
// 1 <= s.length <= 25
// 	s consists of lowercase English letters and parentheses '(' and ')'.
// 	There will be at most 20 parentheses in s.
//

class Solution {
public:
    vector<string> removeInvalidParentheses(string s) {
        unordered_set<string> result;
        int leftRem = 0, rightRem = 0;
        
        // First, determine the number of misplaced left and right parentheses
        for (char c : s) {
            if (c == '(') {
                leftRem++;
            } else if (c == ')') {
                if (leftRem > 0) {
                    leftRem--;
                } else {
                    rightRem++;
                }
            }
        }
        
        backtrack(s, 0, leftRem, rightRem, result);
        
        return vector<string>(result.begin(), result.end());            
    }
};
